### PR TITLE
ceph_volume: support encrypted volumes for lvm new-db/new-wal/migrate  commands

### PR DESCRIPTION
**Pull request by ifed01**
Fixes: https://tracker.ceph.com/issues/55167

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

### CODE REVIEW

**Comment by ifed01** [comment]
@guits - ping.

File: src/ceph-volume/ceph_volume/util/encryption.py [hunk 1]
```diff
@@ -106,17 +106,26 @@ def luks_open(key, device, mapping):
...
     # don't be strict about the remove call, but still warn on the terminal if it fails
+    process.run(['cryptsetup', 'remove', mapping_path], stop_on_error=False)
+
+def dmcrypt_close_by_name(mapping):
+    """
+    Encrypt (close) a device, previously decrypted with cryptsetup
+
+    :param mapping: dmcrypt device name
+    """
+    # don't be strict about the remove call, but still warn on the terminal if it fails
```
Comment (guits) [inline src/ceph-volume/ceph_volume/util/encryption.py:128]: What is the difference between `dmcrypt_close()` and `dmcrypt_close_by_name()` ?

At the end, they both do the same call to `process.run()`

File: src/ceph-volume/ceph_volume/devices/lvm/migrate.py [hunk 2]
```diff
@@ -300,6 +301,15 @@ def get_source_args(self, osd_path, devices):
                 osd_path, self.get_filename_by_type(type))]
         return ret
 
+    def decrypt_close(self, source_devices):
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/lvm/migrate.py:304]: typo?

```suggestion
    def dmcrypt_close(self, source_devices):
```

File: src/ceph-volume/ceph_volume/devices/lvm/migrate.py [hunk 3]
```diff
@@ -340,10 +355,13 @@ def migrate_to_new(self, osd_id, osd_fsid, devices, target_lv):
...
                     'Failed to migrate to : {}'.format(self.args.target))
-            else:
-                system.chown(os.path.join(osd_path, "block.{}".format(
-                    target_type)))
-                terminal.success('Migration successful.')
+
+            system.chown(os.path.join(osd_path, "block.{}".format(
+                target_type)))
+            if tag_tracker.data_device.lv_api.encrypted:
+                self.decrypt_close(source_devices)
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/lvm/migrate.py:362]:
```suggestion
                self.dmcrypt_close(source_devices)
```

File: src/ceph-volume/ceph_volume/devices/lvm/migrate.py [hunk 4]
```diff
@@ -391,8 +409,9 @@ def migrate_to_existing(self, osd_id, osd_fsid, devices, target_lv):
                     'Failed to migrate device, error code:{}'.format(exit_code))
                 raise SystemExit(
                     'Failed to migrate to : {}'.format(self.args.target))
-            else:
-                terminal.success('Migration successful.')
+            if tag_tracker.data_device.lv_api.encrypted:
+                self.decrypt_close(source_devices)
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/lvm/migrate.py:413]:
```suggestion
                self.dmcrypt_close(source_devices)
```

File: src/ceph-volume/ceph_volume/util/encryption.py (hunk 1 above)
Comment (guits) [reply src/ceph-volume/ceph_volume/util/encryption.py:128]: If this is a question passing a mapping name vs. a path, perhaps `dmcrypt_close()` should probably just be reworked a bit.
What do you think?

Comment (ifed01) [reply src/ceph-volume/ceph_volume/util/encryption.py:128]: I can see two points:
1. dmcrypt_close vs. dmcrypt_close_by_name operate on different input parameters: path vs. name. Hence better to distinguish that somehow
2. dmcrypt_close performs additional check on path presence. Additional checks (processing logic) might appear in the future at one of implementations - thus again better to have them different

I'm not very good at Python-style coding so please advice if you think there are more 'pythonish" approaches to implement the above.

Comment (guits) [reply src/ceph-volume/ceph_volume/util/encryption.py:128]: I would probably just adapt the existing function like this instead:


```python
def dmcrypt_close(mapping, skip_path_check=False):
    """
    Encrypt (close) a device, previously decrypted with cryptsetup

    :param mapping:
    """
    if not skip_path_check and not os.path.exists(mapping):
        logger.debug('device mapper path does not exist %s' % mapping)
        logger.debug('will skip cryptsetup removal')
        return
    # don't be strict about the remove call, but still warn on the terminal if it fails
    process.run(['cryptsetup', 'remove', mapping], stop_on_error=stop_on_error)
```
it renames `mapping_path` to `mapping` which, IMO, becomes a bit more generic and adds an argument so you can decide to skip the mapper path check.
(note that it fixes a bug in the line `logger.debug('device mapper path does not exist %s' % mapping)`, `mapping` was undefined so far...)

Comment (ifed01) [reply src/ceph-volume/ceph_volume/util/encryption.py]: @guits - updated as per you suggestion.

File: src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py [hunk 5]
```diff
@@ -990,6 +994,74 @@ def test_newwal_no_systemd(self, is_root, monkeypatch):
             '--dev-target', '/dev/VolGroup/target_volume',
             '--command', 'bluefs-bdev-new-wal']
 
+    @patch('os.getuid')
+    def test_newwal_encrypted(self, m_getuid, monkeypatch, capsys):
+        m_getuid.return_value = 0
```
Comment (guits) [inline src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py:999]:
```suggestion
    def test_newwal_encrypted(self, is_root, monkeypatch, capsys):
```

File: src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py [hunk 6]
```diff
@@ -1245,6 +1326,120 @@ def test_migrate_data_db_to_new_db(self, m_getuid, monkeypatch):
             '--devs-source', '/var/lib/ceph/osd/ceph-2/block',
             '--devs-source', '/var/lib/ceph/osd/ceph-2/block.db']
 
+    @patch('os.getuid')
+    def test_migrate_data_db_to_new_db_encrypted(self, m_getuid, monkeypatch):
+        m_getuid.return_value = 0
```
Comment (guits) [inline src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py:1331]:
```suggestion
    def test_migrate_data_db_to_new_db_encrypted(self, is_root, monkeypatch):
```

File: src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py [hunk 7]
```diff
@@ -1653,6 +1848,147 @@ def test_migrate_data_db_wal_to_new_db(self, m_getuid, monkeypatch):
             '--devs-source', '/var/lib/ceph/osd/ceph-2/block.db',
             '--devs-source', '/var/lib/ceph/osd/ceph-2/block.wal']
 
+    @patch('os.getuid')
+    def test_migrate_data_db_wal_to_new_db_encrypted(self, m_getuid, monkeypatch):
+        m_getuid.return_value = 0
```
Comment (guits) [inline src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py:1853]:
```suggestion
    def test_migrate_data_db_wal_to_new_db_encrypted(self, is_root, monkeypatch):
```

File: src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py [hunk 8]
```diff
@@ -2129,6 +2465,120 @@ def test_migrate_data_wal_to_db(self,
             '--devs-source', '/var/lib/ceph/osd/ceph-2/block',
             '--devs-source', '/var/lib/ceph/osd/ceph-2/block.wal']
 
+    @patch('os.getuid')
+    def test_migrate_data_wal_to_db_encrypted(self,
+                                              m_getuid,
+                                              monkeypatch,
+                                              capsys):
+        m_getuid.return_value = 0
```
Comment (guits) [inline src/ceph-volume/ceph_volume/tests/devices/lvm/test_migrate.py:2473]:
```suggestion
    def test_migrate_data_wal_to_db_encrypted(self,
                                              is_root,
                                              monkeypatch,
                                              capsys):
```

**Comment by guits** [comment]
http://pulpito.ceph.com/gabrioux-2023-07-27_15:32:12-orch:cephadm-wip-guits-testing-5-2023-07-26-0921-distro-default-smithi/

**Review by guits — APPROVED** [review]

File: src/ceph-volume/ceph_volume/devices/lvm/prepare.py [hunk 9]
```diff
@@ -23,19 +23,7 @@ def prepare_dmcrypt(key, device, device_type, tags):
...
-    )
-    encryption_utils.luks_open(
-        key,
-        device,
-        uuid
-    )
-
-    return '/dev/mapper/%s' % uuid
-
+    return encryption_utils.prepare_dmcrypt(key, device, uuid)
```
Comment (guits) [inline src/ceph-volume/ceph_volume/devices/lvm/prepare.py:26]: @ifed01 can we make use of `encryption_utils.prepare_dmcrypt()` in `ceph-volume/ceph_volume/devices/raw/prepare.py` (mainly for consistency...)

Comment (ifed01) [reply src/ceph-volume/ceph_volume/devices/lvm/prepare.py:26]: done

**Comment by ifed01** [comment]
jenkins test make check

**Pull request merged**